src/plugin.js

```diff
@@ -1,96 +1,156 @@
+ */
+function isCompatiblePlugin(obj, propertyName = 'isGitProxyPlugin') {
+  while (obj != null) {
+    if (Object.prototype.hasOwnProperty.call(obj, propertyName) &&
+      obj.isGitProxyPlugin &&
+      Object.keys(obj).includes('exec')) {
+      return true;
+    }
+    obj = Object.getPrototypeOf(obj);
+  }
+  return false;
+}
+
+/**
+ * @typedef PluginTypeResult
+ * @property {ProxyPlugin[]} pushPlugins - List of push plugins
+ * @property {ProxyPlugin[]} pullPlugins - List of pull plugins
+ */
+
+/**
+ * Registers and loads plugins used by git-proxy
  */
 class PluginLoader {
+  /**
+   * @property {Promise} load - A Promise that begins loading plugins from a list of modules. Callers must run `await loader.load` to load plugins.
+   */
+  load;
+  /**
+   * This property is not used in production code. It is exposed for testing purposes.
+   * @property {Promise} ready - A Promise that resolves when all plugins have been loaded.
+   */
+  ready;
   /**
    * Initialize PluginLoader with candidates modules (node_modules or relative
    * file paths).
-   * @param {Array.<string>} names List of Node module/package names to load.
-   * @param {Array.<string>} paths List of file paths to load modules from.
+   * @param {Array.<string>} targets List of Node module package names or files to load.
    */
-  constructor(names, paths) {
-    this.names = names;
-    this.paths = paths;
+  constructor(targets) {
+    this.targets = targets;
     /**
-     * @type {Array.<ProxyPlugin>} List of ProxyPlugin objects loaded.
+     * @type {ProxyPlugin[]} List of loaded ProxyPlugins
      * @public
      */
-    this.plugins = [];
+    this.pushPlugins = [];
+    this.pullPlugins = [];
+    if (this.targets.length === 0) {
+      console.log('No plugins configured'); // TODO: log.debug()
+      this.ready = Promise.resolve();
+      this.load = () => Promise.resolve(); // Ensure this.load is always defined
+      return;
+    }
+    this.load = this._loadPlugins();
   }
 
-  /**
-   * Load configured plugins as modules and set each concrete ProxyPlugin
-   * to this.plugins for use in proxying.
-   */
-  load() {
-    const modulePromises = [];
-    for (const path of this.paths) {
-      modulePromises.push(this._loadFilePlugin(path));
-    }
-    for (const name of this.names) {
-      modulePromises.push(this._loadPackagePlugin(name));
-    }
-    Promise.all(modulePromises).then((vals) => {
-      const modules = vals;
-      console.log(`Found ${modules.length} plugin modules`);
-      const pluginObjPromises = [];
-      for (const mod of modules) {
-        pluginObjPromises.push(this._castToPluginObjects(mod));
+  async _loadPlugins() {
+    try {
+      const modulePromises = this.targets.map(target =>
+        this._loadPluginModule(target).catch(error => {
+          console.error(`Failed to load plugin: ${error}`); // TODO: log.error()
```

Comment:
Another naive comment, but doesn't this risk interpolating something like `[object Object]`?

```suggestion
          console.error(`Failed to load plugin: ,` error); // TODO: log.error()
```

Reply:
`Error` objects should always have a valid toString().

```
$ node
Welcome to Node.js v20.16.0.
Type ".help" for more information.
> let e;
> try { console.log('do something'); throw new Error('message'); } catch (ex) { e = ex }
do something
Error: message
    at REPL8:1:42
    at ContextifyScript.runInThisContext (node:vm:136:12)
    at REPLServer.defaultEval (node:repl:598:22)
    at bound (node:domain:432:15)
    at REPLServer.runBound [as eval] (node:domain:443:12)
    at REPLServer.onLine (node:repl:927:10)
    at REPLServer.emit (node:events:531:35)
    at REPLServer.emit (node:domain:488:12)
    at [_onLine] [as _onLine] (node:internal/readline/interface:417:12)
    at [_line] [as _line] (node:internal/readline/interface:888:18)
> e
Error: message
    at REPL8:1:42
    at ContextifyScript.runInThisContext (node:vm:136:12)
    at REPLServer.defaultEval (node:repl:598:22)
    at bound (node:domain:432:15)
    at REPLServer.runBound [as eval] (node:domain:443:12)
    at REPLServer.onLine (node:repl:927:10)
    at REPLServer.emit (node:events:531:35)
    at REPLServer.emit (node:domain:488:12)
    at [_onLine] [as _onLine] (node:internal/readline/interface:417:12)
    at [_line] [as _line] (node:internal/readline/interface:888:18)
> console.log(`an error was received: ${e}`)
an error was received: Error: message
undefined
```